fs: include more fs.stat*() optimizations

mscdex opened the pull request:
`stat*()` optimizations including:

* Move async `*stat()` functions to `FillStatsArray()` now used by the sync `*stat()` functions

* Avoid creating `fs.Stats` instances for implicit async/sync `*stat()` calls used in various `fs` functions

* Store reference to `Float64Array` data on C++ side for easier/faster access, instead of passing from JS to C++ on every async/sync `*stat()` call

Additionally, there are some further `realpath*()` optimizations, including:

* Skip URL instance check for common (string) cases

* Avoid regexp on non-Windows platforms when parsing the root of a path

* Skip call to `getOptions()` in common case where no `options` is passed

* Avoid `hasOwnProperty()`

Some benchmark results with all of the changes in this PR:

```
                                                                        improvement confidence      p.value
 fs/readfile.js concurrent=1 len=1024 dur=5                                  5.50 %            0.177524964
 fs/readfile.js concurrent=1 len=16777216 dur=5                             -0.26 %            0.860522023
 fs/readfile.js concurrent=10 len=1024 dur=5                                 7.57 %         ** 0.004268331
 fs/readfile.js concurrent=10 len=16777216 dur=5                            -0.35 %            0.374914593
 fs/bench-stat.js kind="fstat" n=200000                                      3.71 %         ** 0.002509243
 fs/bench-stat.js kind="lstat" n=200000                                      3.16 %         ** 0.001381068
 fs/bench-stat.js kind="stat" n=200000                                       4.13 %          * 0.019328676
 fs/readFileSync.js n=600000                                                20.27 %        *** 2.313943e-38
 fs/bench-statSync.js kind="fstatSync" n=1000000                             2.01 %            1.066663e-01
 fs/bench-statSync.js kind="lstatSync" n=1000000                             4.42 %        *** 4.163876e-06
 fs/bench-statSync.js kind="statSync" n=1000000                              3.46 %        *** 1.333101e-06
 fs/bench-realpathSync.js type="resolved" n=300000                          86.06 %        *** 3.681945e-42
 fs/bench-realpathSync.js type="relative" n=400000                          61.71 %        *** 1.349012e-54
 fs/bench-realpath.js type="resolved" n=30000                                6.00 %        *** 3.601045e-11
 fs/bench-realpath.js type="relative" n=100000                               7.61 %        *** 3.163175e-14
 module/module-loader.js useCache="false" fullPath="false" thousands=50      5.05 %        *** 1.426385e-19
 module/module-loader.js useCache="false" fullPath="true" thousands=50       5.15 %        *** 6.171543e-16
```

I'm initially marking this as semver-major because of the changes in `fs.readFile()`, `fs.readFileSync()`, etc. in how they retrieve stats for files/directories. If someone monkey patches `fs.*sync*()`, those monkey matched methods would no longer be called from those changed functions (`fs.readFile()`, etc.). If we feel we shouldn't be worried about this, then I will gladly remove the label.

CI: https://ci.nodejs.org/job/node-test-pull-request/6674/

##### Checklist

- [X] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [X] tests and/or benchmarks are included
- [X] commit message follows commit guidelines

##### Affected core subsystem(s)

* fs

not-an-aardvark commented:
For what it's worth, this will break some modules like [mock-fs](https://github.com/tschaub/mock-fs) which rely on mutating `process.binding('fs')` and shimming everything to avoid touching the actual filesystem.

I acknowledge that `process.binding('fs')` is not a public API, so changes like this might be worthwhile anyway. I just want to point out that changes to the API of `process.binding('fs')` methods do currently have ecosystem impact.

mscdex commented:
@not-an-aardvark Well, in a recent PR I was told not to worry about `process.binding()` changes \*shrug\*. See [this comment](https://github.com/nodejs/node/pull/11522#issuecomment-282101101) and [this comment](https://github.com/nodejs/node/pull/11522#issuecomment-282164291).

not-an-aardvark commented:
Well, the reason I bring this up is that https://github.com/nodejs/node/pull/11522 ended up causing https://github.com/tschaub/mock-fs/issues/197 and breaking some builds. (I'm not trying to argue that this particular decision was justified/unjustified -- I just want to make sure that these decisions are made with full information on what they might break downstream.)

jasnell commented on lib/fs.js:
```diff
@@ -130,7 +149,7 @@ function isFd(path) {
 
-// Static method to set the stats properties on a Stats object.
+// Constructor for file stats
 function Stats(
```
btw, it's worth noting that I had tried to convert this to a ES6 Class just to see what would happen and it yielded a 50% perf *reduction*.

mscdex commented:
Updated.

addaleax commented on lib/fs.js:
```diff
@@ -1369,13 +1408,13 @@ function StatWatcher() {
     if (oldStatus === -1 &&
         newStatus === -1 &&
-        current.nlink === previous.nlink) return;
+        statValues[2] === statValues[16]) return;
```
It might be good to leave a comment to indicate that these values refer to `.nlink`

bnoordhuis replied:
In another comment I suggested adding constants for the field indices, that would help with legibility.

mscdex replied:
Changed here as well now.

mscdex commented:
CI again after making suggested changes: https://ci.nodejs.org/job/node-test-pull-request/6719/

bnoordhuis commented:
What I had in mind was more along the lines of `const kInoFieldIndex = ...` declarations at the top of the file and use those consistently everywhere, with a `const kStatFieldCount = 14` so you can use `statValues[kStatFieldCount + kInoFieldIndex]` to index into the second stat values list.

mscdex commented:
@bnoordhuis I was trying to avoid lookups where possible as I'm not sure V8 will always inline the value, even if `const`.

bnoordhuis commented:
Valid concern.  One more argument for reintroducing macros, I suppose.  Alternatively, you could write them as `statValues[/* ino */ 7]`.

Fishrock123 commented:
> For what it's worth, this will break some modules like [mock-fs](https://github.com/tschaub/mock-fs) which rely on mutating `process.binding('fs')` and shimming everything to avoid touching the actual filesystem.
>
> I acknowledge that `process.binding('fs')` is not a public API, so changes like this might be worthwhile anyway. I just want to point out that changes to the API of `process.binding('fs')` methods do currently have ecosystem impact.

We should warn them ahead of time still I think. @mscdex do you figure the new way is still shimmable?

mscdex commented:
@Fishrock123 Support for the `Object::New()`/`BuildStatsObject()` code paths for `*statSync()` was already removed in https://github.com/nodejs/node/pull/11522 because it was decided we aren't supporting `process.binding()` changes.

mscdex commented:
Rebased and added inline comments next to the Float64Array's indices.

mscdex commented:
Alright, I've fixed the botched rebase I previously did.

I have also now included some more `realpath*()` optimizations and added relevant benchmark results to the original post.

CI: https://ci.nodejs.org/job/node-test-pull-request/6803/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/640/

mscdex commented:
Alright, I've fixed the final `graceful-fs`-related issue.

CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/646/

No more related CITGM failures!

CI again: https://ci.nodejs.org/job/node-test-pull-request/6818/

mscdex commented:
CI is green. I'd like some re-reviews of the recent changes if possible.

mcollina commented:
Still LGTM, good work @mscdex

tschaub commented:
> Fair enough. The reason I bring this up is that #11522 ended up causing tschaub/mock-fs#197 and breaking some builds. (I'm not trying to argue that this particular decision was justified/unjustified -- I just want to make sure that these decisions are made with full information on what they might break downstream.)

Thanks for bringing this up @not-an-aardvark - and for providing a fix.

Off topic for this PR I know, but I thought it would be worth mentioning that `binding('fs')` has proven to be a more stable interface (or at least easier to implement in a backwards compatible way) than the `fs` module itself  based on personal experience having implemented it from Node 0.8 through Node 7.

Clearly others shouldn't rely on `binding('fs')` being stable (particularly between major releases).  I just wanted to point out that historically it has been more "stable" than the `fs` module.

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/6849/

jasnell commented:
Landed in 6a5ab5d...7109774

kunalspathak commented on src/node_file.cc:
```diff
@@ -1496,12 +1377,20 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+    fields = new double[2 * 14];
+    env->set_fs_stats_field_array(fields);
+  }
+  Local<ArrayBuffer> ab = ArrayBuffer::New(env->isolate(),
```
@mscdex , Correct me if i am wrong, but `ArrayBuffer::New()` creates a buffer by default in [externalizeMode](https://github.com/nodejs/node/blob/master/deps/v8/include/v8.h#L4017-L4020). Who frees `fields` array in that case? May be you want to pass `ArrayBufferCreationMode::kExternalized` to this API?

mscdex replied:
That's the [default mode](https://v8docs.nodesource.com/node-7.4/d5/d6e/classv8_1_1_array_buffer.html#a98c19b0bf91cc3ea22371fbab5707e25). Also the description says:

> Create a new **ArrayBuffer** over an existing memory block. The created array buffer is by default immediately in externalized state. The memory block will not be reclaimed when a created **ArrayBuffer** is garbage-collected.

kunalspathak replied:
Right, but then how/when do we delete `fields`?

kunalspathak replied:
> May be you want to pass ArrayBufferCreationMode::kExternalized to this API?

What i meant was to pass `ArrayBufferCreationMode::kInternalized` instead.

mscdex replied:
We don't. It's created once and stays for the life of the process, for whenever someone calls the `fs.stat*()` functions implicitly or explicitly.

kunalspathak replied:
Got it.

mscdex commented:
@nodejs/ctc How do you feel about the semver-major-ness of these changes? Do you think it is warranted? The reason I ask is that apparently in some situations certain `fs.Stats` fields are negative in JS land because the previous `stat()` functions were converting unsigned integers to signed integers (see https://github.com/nodejs/node/issues/12419).

So either we could decide this PR really isn't semver-major and backport it as-is to v7.x which still has async `stat()` calls returning `fs.Stats` created using the old method, *or* I extract the async `stat()` changes from this PR separately and only backport that to v7.x. The downside to the latter solution is that it might add confusion due to the partial backport-ness.

Thoughts?

addaleax commented:
You mean, using this as a bugfix for https://github.com/nodejs/node/issues/12419? I would be okay with that, but it’s probably also worth mentioning that https://github.com/nodejs/node/pull/11522 made it into v6.10.1, so for being consistent we should also apply it there (or, alternatively, revert https://github.com/nodejs/node/pull/11522 on v6.x).

mscdex commented:
@addaleax Ah ok I did not know it made it into v6.x.

jasnell commented:
If there have been no regressions I'm good with downgrading thing to semver-minor and backporting.

mscdex commented:
@jasnell Have enough people tested with master though since this was landed?

jasnell commented:
Unsure. Preferably it would sit through at least one release on current before backporting.

mscdex commented:
I think we'd need to make a decision sooner than that because of the issue I linked to.

jasnell commented:
Doh, actually nevermind lol. For some reason I was reading that as backporting to v6. I'm good with backporting to v7.

mscdex commented:
Well it's the same situation for v6 really, unless we want to decide differently for that branch.

jasnell commented:
Yep, understood. Let's get it into a v7 release. Give it a week or two, then decide for v6.
